Source/JavaScriptCore/runtime/PropertyTable.h

```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
+        return findImpl(index, tableFromIndexVector(index), key);
+    }
+    const auto* index = bitwise_cast<const uint32_t*>(indexVector & indexVectorMask);
+    return findImpl(index, tableFromIndexVector(index), key);
+}
+
+inline std::tuple<PropertyOffset, unsigned> PropertyTable::get(const KeyType& key)
 {
     ASSERT(key);
     ASSERT(key->isAtom() || key->isSymbol());
     ASSERT(key != PROPERTY_MAP_DELETED_ENTRY_KEY);
 
     if (!m_keyCount)
-        return nullptr;
-
-    unsigned hash = IdentifierRepHash::hash(key);
-
-#if DUMP_PROPERTYMAP_STATS
-    ++propertyTableStats->numLookups;
-#endif
-
-    while (true) {
-        unsigned entryIndex = m_index[hash & m_indexMask];
-        if (entryIndex == EmptyEntryIndex)
-            return nullptr;
-        if (key == table()[entryIndex - 1].key) {
-            ASSERT(!m_deletedOffsets || !m_deletedOffsets->contains(table()[entryIndex - 1].offset));
-            return &table()[entryIndex - 1];
-        }
-
-#if DUMP_PROPERTYMAP_STATS
-        ++propertyTableStats->numLookupProbing;
-#endif
+        return std::tuple { invalidOffset, 0 };
 
-        hash++;
-    }
+    auto [entryIndex, index, offset, attributes] = find(key);
+    UNUSED_VARIABLE(entryIndex);
+    UNUSED_VARIABLE(index);
+    return std::tuple { offset, attributes };
 }
 
-inline std::pair<PropertyTable::find_iterator, bool> WARN_UNUSED_RETURN PropertyTable::add(VM& vm, const ValueType& entry)
+inline std::tuple<PropertyOffset, unsigned, bool> WARN_UNUSED_RETURN PropertyTable::add(VM& vm, const ValueType& entry)
 {
-    ASSERT(!m_deletedOffsets || !m_deletedOffsets->contains(entry.offset));
+    ASSERT(!m_deletedOffsets || !m_deletedOffsets->contains(entry.offset()));
 
     // Look for a value with a matching key already in the array.
-    find_iterator iter = find(entry.key);
-    if (iter.first)
-        return std::make_pair(iter, false);
+    auto [entryIndex, index, offset, attributes] = find(entry.key());
+    if (offset != invalidOffset)
+        return std::tuple { offset, attributes, false };
 
 #if DUMP_PROPERTYMAP_STATS
     ++propertyTableStats->numAdds;
 #endif
 
     // Ref the key
-    entry.key->ref();
+    entry.key()->ref();
 
     // ensure capacity is available.
-    if (!canInsert()) {
-        rehash(vm, m_keyCount + 1);
-        iter = find(entry.key);
-        ASSERT(!iter.first);
+    if (!canInsert(entry)) {
+        rehash(vm, m_keyCount + 1, !canBeFitInCompact(entry));
+        std::tie(entryIndex, index, offset, attributes) = find(entry.key());
+        ASSERT_UNUSED(offset, offset == invalidOffset);
+        ASSERT_UNUSED(entryIndex, entryIndex == EmptyEntryIndex);
     }
 
     // Allocate a slot in the hashtable, and set the index to reference this.
-    unsigned entryIndex = usedCount() + 1;
-    m_index[iter.second] = entryIndex;
-    iter.first = &table()[entryIndex - 1];
-    *iter.first = entry;
+    entryIndex = usedCount() + 1;
```

Comment:
in the isCompact case, can we assert `usedCount() < UINT8_MAX`? I also think entryIndex here is `unsigned`, so alternatively, assert that it's <= UINT8_MAX. Since this works by canInsert check above.